featurebyte/service/base_table_validation.py

```diff
@@ -87,9 +96,65 @@ def table_needs_validation(cls, table_model: Document) -> bool:
         -------
         bool
         """
-        _ = table_model
+        assert isinstance(table_model, TableModel)
+        for col_info in table_model.columns_info:
+            if (
+                col_info.dtype == DBVarType.VARCHAR
+                and col_info.dtype_metadata
+                and col_info.dtype_metadata.timestamp_schema
+            ):
+                return True
         return False
 
+    @staticmethod
+    async def _validate_timestamp_format_string(
+        col_info: ColumnInfo, session: BaseSession, table_model: TableModel, num_records: int
+    ) -> None:
+        assert col_info.dtype_metadata is not None
+        assert col_info.dtype_metadata.timestamp_schema is not None
+
+        adapter = session.adapter
+        timestamp_schema = col_info.dtype_metadata.timestamp_schema
+        assert timestamp_schema.format_string is not None
+
+        query_expr = (
+            select(
+                expressions.alias_(
+                    adapter.to_timestamp_from_string(
+                        quoted_identifier(col_info.name),
+                        timestamp_schema.format_string,
+                    ),
+                    alias=col_info.name,
+                    quoted=True,
+                )
+            )
+            .from_(
+                get_fully_qualified_table_name(
+                    table_model.tabular_source.table_details.model_dump()
+                )
+            )
+            .limit(num_records)
+        )
+        query = sql_to_string(query_expr, source_type=adapter.source_type)
+        # check that the format string is valid for the first num_records
+        try:
+            await session.execute_query_long_running(query)
+        except Exception as exc:
+            raise TableValidationError(
+                f"Timestamp column '{col_info.name}' has invalid format string '{timestamp_schema.format_string}'. "
+                f"Error: {str(exc)}"
+            )
+
+        # check no more than one timezone information
+        if (
+            adapter.format_string_has_timezone(timestamp_schema.format_string)
+            and timestamp_schema.timezone is not None
+        ):
+            raise TableValidationError(
+                f"Timestamp column '{col_info.name}' has timezone information in the data and in the schema. "
+                f"Please remove timezone information from the data or from the schema."
+            )
```

Comment:
Would it make sense to perform this check earlier during table registration synchronously since it doesn't require executing any database queries?

Reply:
Good point. Will move this to table registration.

Reply:
If we want to perform this check during table registration, that means we need to create a session object during table registration. Just want to check whether this is ok (as we may see long time session object creation sometime).

Reply:
Maybe we should construct adapter directly without constructing the session object. 🤔